server/app/services/applicant/ApplicantServiceImpl.java

```diff
@@ -358,6 +367,138 @@ public ImmutableList<Application> getAllApplications() {
+            .toCompletableFuture();
+    ImmutableList<ProgramDefinition> activePrograms =
+        versionRepository.getActiveVersion().getPrograms().stream()
+            .map(Program::getProgramDefinition)
+            .filter(pdef -> pdef.displayMode().equals(DisplayMode.PUBLIC))
+            .collect(ImmutableList.toImmutableList());
+
+    return applicationsFuture.thenApplyAsync(
+        applications -> {
+          checkForDuplicateDrafts(applications);
+          return relevantProgramsForApplicant(activePrograms, applications);
+        },
+        httpExecutionContext.current());
+  }
+
+  private RelevantPrograms relevantProgramsForApplicant(
+      ImmutableList<ProgramDefinition> activePrograms, ImmutableSet<Application> applications) {
+    ImmutableMap<String, ProgramDefinition> activeProgramNames =
+        ImmutableMap.copyOf(
+            activePrograms.stream()
+                .collect(Collectors.toMap(ProgramDefinition::adminName, pdef -> pdef)));
+
+    // When new versions of Programs are created, they have distinct IDs but retain the
+    // same adminName. In order to find the most recent draft / active application,
+    // we first group by the unique name rather than the ID.
+    Map<String, Map<LifecycleStage, Optional<Application>>> mostRecentApplicationsByProgram =
+        applications.stream()
+            .collect(
+                Collectors.groupingBy(
+                    a -> {
+                      return a.getProgram().getProgramDefinition().adminName();
+                    },
+                    Collectors.groupingBy(
+                        Application::getLifecycleStage,
+                        // In practice, we don't expect an applicant to have multiple applications
+                        // for a given program / LifecycleStage combination. Grabbing the latest
+                        // application here guards against that case, should it occur.
+                        Collectors.maxBy(Comparator.comparingLong(a -> a.id)))));
+
+    ImmutableList.Builder<ApplicantProgramData> inProgressPrograms = ImmutableList.builder();
+    ImmutableList.Builder<ApplicantProgramData> submittedPrograms = ImmutableList.builder();
+    ImmutableList.Builder<ApplicantProgramData> unappliedPrograms = ImmutableList.builder();
+
+    mostRecentApplicationsByProgram.forEach(
+        (programName, appByStage) -> {
+          Optional<Application> maybeDraftApp =
+              appByStage.getOrDefault(LifecycleStage.DRAFT, Optional.empty());
+          Optional<Application> maybeSubmittedApp =
+              appByStage.getOrDefault(LifecycleStage.ACTIVE, Optional.empty());
+          Optional<Instant> maybeSubmitTime = maybeSubmittedApp.map(Application::getSubmitTime);
+          if (maybeDraftApp.isPresent()) {
+            // We want to ensure that any generated links points to the version
+            // of the program associated with the draft, not the most recent version.
```

Comment:
I don't think this comment is necessary, the code seems fairly obvious and readable without restating how the system modeling works. 

if you keep the comment be clearer about lifecyclstage / table since stages are used in both modelings. eg "draft application" "most recent program version"

Reply:
Removed the comment.